Design/Fit functions in python.md

```diff
@@ -0,0 +1,198 @@
+```
+
+Tying to a constant value that a parameter currently has can also be done with the `fix` method.
+```
+  func.fix('A0')
+  func.fix('f1.A2', 'f2.A2')
+```
+
+Ties are removed with the `removeTie` or `free` methods.
+```
+  func.free('f1.Sigma', 'f3.Sigma')
+  func.removeTie('f1.Sigma', 'f3.Sigma')
+```
+
+Composite functions should be able to set ties on all members with a single call:
+```
+  c.tieAll('Sigma')
+```
+The above call should tie all `Sigma` parameters in the member functions. It is equivalent to calling
+```
+  c.tie({'f1.Sigma': 'f0.Sigma', 'f2.Sigma': 'f0.Sigma', 'f3.Sigma': 'f0.Sigma', ...})
+```
+Similarly a call to `fixAll` fixes all parameters with the given name to their current value.
+```
+  spectrum.fixAll('FWHM')
+  spectrum.fixAll('f1.FWHM': 0.01)
+```
+Calling `fixAll()`/`freeAll()` without arguments fixes/frees all parameters of a function.
+
+### Setting constraints
+
+Constraints are set with `constraints` method passing constraints as strings:
+```
+  g.constraints('-1 < PeakCentre < 1', 'Sigma > 0')
+```
+
+For composite functions constraints can be set both through the composite function and its members:
+```
+  c = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+  c.constraints('f1.Sigma > 0')
+  c[2].constraints('0 < Sigma < 1')
+```
+
+Composite functions should be able to set constraints on all members with a single call:
+```
+  c.constrainAll('Height > 1', 'Sigma > 0')
+```
+This should constrain the named parameters in all members that have them. Members that don't have these parameters should be skipped.
+
+## Fitting
+
+Python function `Fit` should accept instances of `FitFunctionWrapper` as its `Function` argument. `Fit` does't modify the input function but the returned value has the output function with the fitted parameters. The namedtuple returned by `Fit` must contain the fitted function:
+```
+  res = Fit(func, ws)
+  output_func = res.Function
+```
+
+## Function evaluation
+
+`FitFunctionWrapper` class will implement `__call__(...)` method to evaluate the function. The method will be able to accept a variety of input types:
+ 1. Workspace
+ 2. Numpy array
+ 3. List of numbers
+ 4. Single number
+ 
+The output type is the same as that of the input. For example:
+```
+  g = Gaussian(Sigma=0.1, Height=1)
+  y1 = g(0.2)
+  y2 = g(2*g.Sigma)
+  y_list = g([-2, -1, 0, 1, 2])
+  out_ws = g(ws)
+```
+
+## Plotting
+
+Python function `plotFunction` will plot one or more fitting functions in MantidPlot. It will have the arguments:
+ 1. first positional argument - the function(s) to plot
+ 2. `workspace` - optional workspace on which x-values to evaluate the function
+ 3. `workspaceIndex` - optional index of a spectrum to get the x-values from
+ 4. `startX` - optional if `workspace` is given, mandatory otherwise. The start of the plot region.
+ 5. `endX` - optional if `workspace` is given, mandatory otherwise. The end of the plot region.
```

Comment:
regarding plotting, having a python method is more intuitive to me than an external function.
```
f=Lorentzian()
f.plot(*args, **args) # workspace, workspaceindex, startX, endX
```
Sometimes the workspace and workspaceIndex have been already passed with the SetWorkspace method.
```
f=TeixeiraWaterSQE()
f.setMatrixWorkspace(workspace, workspaceIndex) # determines attribute Q
f.plot(startX, endX)
```
or you could have both :)
```
def plot(self, *args, **kwargs):
  plotFunction(self, *args, **kwargs)
```